subgraphs/tornado-cash/src/common/metrics.ts

```diff
@@ -0,0 +1,219 @@
+
+  usageMetricsHourly.blockNumber = event.block.number;
+  usageMetricsHourly.timestamp = event.block.timestamp;
+  usageMetricsHourly.hourlyTransactionCount += INT_ONE;
+
+  // Number of days since Unix epoch
+  let day = event.block.timestamp.toI32() / SECONDS_PER_DAY;
+  let hour = event.block.timestamp.toI32() / SECONDS_PER_HOUR;
+
+  let dayId = day.toString();
+  let hourId = hour.toString();
+
+  // Combine the id and the user address to generate a unique user id for the day
+  let dailyActiveAccountId = from.concat("-").concat(dayId);
+  let dailyActiveAccount = ActiveAccount.load(dailyActiveAccountId);
+  if (!dailyActiveAccount) {
+    dailyActiveAccount = new ActiveAccount(dailyActiveAccountId);
+    usageMetricsDaily.dailyActiveUsers += INT_ONE;
+    dailyActiveAccount.save();
+  }
+
+  let hourlyActiveAccountId = from.concat("-").concat(hourId);
+  let hourlyActiveAccount = ActiveAccount.load(hourlyActiveAccountId);
+  if (!hourlyActiveAccount) {
+    hourlyActiveAccount = new ActiveAccount(hourlyActiveAccountId);
+    usageMetricsHourly.hourlyActiveUsers += INT_ONE;
+    hourlyActiveAccount.save();
+  }
+
+  let account = Account.load(from);
+  if (!account) {
+    account = new Account(from);
+    protocol.cumulativeUniqueUsers += INT_ONE;
+    account.save();
+  }
+  usageMetricsDaily.cumulativeUniqueUsers = protocol.cumulativeUniqueUsers;
+  usageMetricsHourly.cumulativeUniqueUsers = protocol.cumulativeUniqueUsers;
+
+  usageMetricsDaily.totalPoolCount = protocol.totalPoolCount;
+
+  usageMetricsDaily.save();
+  usageMetricsHourly.save();
+  protocol.save();
+}
+
+// Update FinancialsDailySnapshots entity
+export function updateFinancials(event: ethereum.Event): void {
+  let financialMetricsDaily = getOrCreateFinancialsDailySnapshot(event);
+
+  let protocol = getOrCreateProtocol();
+
+  let pools = protocol.pools;
+  let tvl = BIGDECIMAL_ZERO;
+  for (let i = 0; i < pools.length; i++) {
+    let pool = getOrCreatePool(pools[i], event);
+
+    tvl = tvl.plus(pool.totalValueLockedUSD);
+  }
+  protocol.totalValueLockedUSD = tvl;
+
+  financialMetricsDaily.blockNumber = event.block.number;
+  financialMetricsDaily.timestamp = event.block.timestamp;
+  financialMetricsDaily.totalValueLockedUSD = protocol.totalValueLockedUSD;
+  financialMetricsDaily.cumulativeSupplySideRevenueUSD =
+    protocol.cumulativeSupplySideRevenueUSD;
+  financialMetricsDaily.cumulativeProtocolSideRevenueUSD =
+    protocol.cumulativeProtocolSideRevenueUSD;
+  financialMetricsDaily.cumulativeTotalRevenueUSD =
+    protocol.cumulativeTotalRevenueUSD;
+
+  financialMetricsDaily.save();
+  protocol.save();
+}
+
+export function updateRevenue(
+  event: ethereum.Event,
+  poolAddress: string,
+  relayerFeeUsd: BigDecimal,
+  protocolFeeUsd: BigDecimal
+): void {
+  let pool = getOrCreatePool(poolAddress, event);
+
+  pool.cumulativeTotalRevenueUSD = pool.cumulativeTotalRevenueUSD.plus(
+    relayerFeeUsd
+  );
```

Comment:
Total revenue shouldn't include both relayer and protocol fees?

Reply:
@SerenityFund could you please take up questions on revenue calculation?

Reply:
It shouldn't, protocol fee is generally accounted for in relayer fee. Typically, `relayerFee` is paid to the relayer to cover transaction cost + protocol fee + relayer's profit.

Reply:
Understood